tiledb/sm/filesystem/s3.h

```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+  }
+
+ private:
+  /**
+   * Advance to the next object accepted by the filters for this scan.
+   *
+   * @param ptr Reference to the current data pointer.
+   * @sa LsScanIterator::operator++()
+   */
+  void next(typename Iterator::pointer& ptr);
+
+  /**
+   * Advance the iterator to the next object in the current batch of results.
+   * If the iterator is at the end of the current batch, this will fetch the
+   * next batch of results from S3. This does not check if the results are
+   * accepted by the filters for this scan.
+   *
+   * @param ptr Reference to the current data pointer.
+   */
+  void advance(typename Iterator::pointer& ptr) {
+    ++ptr;
+    if (ptr == end_) {
+      if (more_to_fetch()) {
+        // Fetch results and reset the iterator.
+        ptr = fetch_results();
+      } else {
+        // Set the pointer to nullptr to indicate the end of results.
+        end_ = ptr = nullptr;
+      }
+    }
+  }
+
+  /**
+   * Fetch the next batch of results from S3. This also handles setting the
+   * continuation token for the next request, if the results were truncated.
+   *
+   * @return A pointer to the first result in the new batch. The return value
+   *    is used to update the pointer managed by the iterator during traversal.
+   *    If the request returned no results, this will return nullptr to mark the
+   *    end of the scan.
+   * @sa LsScanIterator::operator++()
+   * @sa S3Scanner::next(typename Iterator::pointer&)
+   */
+  typename Iterator::pointer fetch_results() {
+    // If this is our first request, GetIsTruncated() will be false.
+    if (more_to_fetch()) {
+      // If results are truncated on a subsequent request, we set the next
+      // continuation token before resubmitting our request.
+      Aws::String next_marker =
+          list_objects_outcome_.GetResult().GetNextContinuationToken();
+      if (next_marker.empty()) {
+        throw S3Exception(
+            "Failed to retrieve next continuation token for ListObjectsV2 "
+            "request.");
+      }
+      list_objects_request_.SetContinuationToken(std::move(next_marker));
+    }
+    list_objects_outcome_ = client_->ListObjectsV2(list_objects_request_);
+    if (!list_objects_outcome_.IsSuccess()) {
+      throw S3Exception(
+          std::string("Error while listing with prefix '") +
+          this->prefix_.add_trailing_slash().to_string() + "' and delimiter '" +
+          delimiter_ + "'" + outcome_error_message(list_objects_outcome_));
+    }
+    // Update pointers to the newly fetched results.
+    begin_ = &list_objects_outcome_.GetResult().GetContents().front();
+    end_ = &list_objects_outcome_.GetResult().GetContents().back() + 1;
+
+    if (list_objects_outcome_.GetResult().GetContents().empty()) {
+      // If the request returned no results, we've reached the end of the scan.
+      // We hit this case when the number of objects in the bucket is a multiple
+      // of the current max_keys.
+      return nullptr;
```

Comment:
Would be better to explicitly return an end-iterator here. This code assumes knowledge of how the end-iterator is constructed.